contributors/design-proposals/node/troubleshoot-running-pods.md

```diff
@@ -106,13 +100,91 @@ subsequently be used to reattach and is reported by `kubectl describe`.
 
 ### Kubernetes API Changes
 
-#### Chosen Solution: "exec++"
+There has been much discussion about how this fits best into the Kubernetes API.
+The consensus is for an imperative "debug this pod" action whereby the kubelet
+creates a new, temporary container in a pod on command. SIG Node would like to
+avoid new dependencies in the kubelet, so this will be implemented in the Core
+API. Three possible implementations follow, and additional implementations that
+were evaluated and dismissed are at the end of this document.
+
+All of the proposed solutions implement the user-level concept of a _Debug
+Container_ using the API-level concept of an _Ephemeral Container_. The API
+doesn't prescribe how an Ephemeral Container is used. It could conceivably see
+use other than Debug Containers, but we don't currently have other use cases.
+
+#### Chosen Solution: Subresource to Update PodStatus
+
+An Ephemeral Container is not part of the pod specification as it's not part of
+the declared state of the pod, but we describe it using the same primitives as
+in `PodSpec`. An `EphemeralContainer` contains a Spec, a Status and a Target:
+
+```
+// EphemeralContainer describes a container to attach to a running pod for troubleshooting.
+type EphemeralContainer struct {
+        metav1.TypeMeta `json:",inline"`
+
+        // Spec describes the Ephemeral Container to be created.
+        Spec *Container `json:"spec,omitempty" protobuf:"bytes,2,opt,name=spec"`
+
+        // Most recently observed status of the container.
+        // This data may not be up to date.
+        // Populated by the system.
+        // Read-only.
+        // +optional
+        Status *ContainerStatus `json:"status,omitempty" protobuf:"bytes,3,opt,name=status"`
```

Comment:
I'm second guessing my decision to embed the `ContainerStatus` into `EphemeralContainer` as I'm polishing up the kubelet implementation where things just expect to see a `[]ContainerStatus`.  I wonder if it would be better to do something like:

```
type EphemeralContainer struct {
    ...
    Spec *Container
    TargetContainerName string
}

type PodStatus struct {
   ...
   EphemeralContainers []EphemeralContainer
   EphemeralContainerStatuses []ContainerStatus
}
```

I guess that's just pushing the complexity of merging by name onto the client, though, where presumably the spec & status will frequently be operated upon in tandem. If targeting were unnecessary (or could be specified in `Container`) then we wouldn't need the `type EphemeralContainer` at all.

Do you have any suggestions @yujuhong or @liggitt ?

Reply:
I think the originally proposed `EphemeralContainer` struct that groups everything together might be slightly better to parse.